[NVPTX] Add float to tf32 conversion intrinsics

durga4github opened the pull request:
This patch adds the missing variants of float to tf32 conversion
intrinsics, with their corresponding lit tests.

PTX Spec link:
https://docs.nvidia.com/cuda/parallel-thread-execution/#data-movement-and-conversion-instructions-cvt

durga4github commented:
@Artem-B , could you please help with a review?

durga4github commented:
cc: @LewisCrawford

AlexMaclean commented on llvm/include/llvm/IR/IntrinsicsNVVM.td:
```diff
@@ -1466,6 +1466,15 @@ let TargetPrefix = "nvvm" in {
+// Convert Float to TF32
+def int_nvvm_cvt_float_to_tf32 : Intrinsic<[llvm_i32_ty],
+    [llvm_float_ty, // Input float
+     llvm_i8_ty,    // Flag for Rounding Modes
```
Can the rounding mode be expressed with metadata as is done with the existing constrained floating point intrinsic? Expressing enums with many values as ints can make things more difficult to read.

durga4github replied:
Yes, it makes sense. Updated the rounding modes to use metadata strings in the latest revision.

AlexMaclean replied:
Thanks. cc @andykaylor (original author of constrained fp intrinsics) for any thoughts on if this is still the best way to represent this? I think operand bundles were discussed at one point.

andykaylor replied:
Yeah, there are a few patches in progress now to convert things to use operand bundles. The motivation there is to be able to attach constrained behavior to any intrinsic without requiring a constrained-specific form of the intrinsic.

The use of rounding mode here is a bit different than in the constrained intrinsics. In the constrained intrinsics, the rounding mode is meant to indicate the rounding mode that may be assumed. While it is permissable to lower the constrained intrinsics to instructions that directly encode the rounding mode, that's not required. Whereas here, it seems that the rounding mode is meant to be more active such that it is required to be encoded in the instruction.

In several other target-specific intrinsics that include an explicit rounding mode, the rounding mode is specified as an immediate argument, but I think that's usually because it corresponds to a value in the target instruction encoding. Since you need a rounding more printer that performs a translation here, I don't think there's any advantage to using an immediate, which would be a bit of a magic number in the IR.

Artem-B replied:
Naming, the hardest problem in computer science...

My usual position is that I prefer to avoid parameters as instruction selectors, unless it is absolutely necessary, and it does not seem to be the case here.
The main reason is that there's no good way to handle invalid selector values. If someone gets intrinsic name wrong, they do get a sensible diagnostic that the name is wong. With the wrong values, we'll either crash somewhere in LLVM because we can't lower the intrinsic, or will produce something invalid.
In many cases, the set of allowed variants is not a complete cartesian product of all parameters. There are usually invalid combinations. Exact enumeration goes a long way to avoid accidental use of those. It also makes it very obvious which combinations are supported/valid.

In other words, I would prefer to keep things simple, add a handful of new conversion intrinsics mapping 1:1 to the instructions, and be done with that.

durga4github replied:
I thought your suggestion was to have these modifiers as flags, from the discussion here:
https://github.com/llvm/llvm-project/pull/118965#discussion_r1884417829

We have handled the invalid cases here with error messages from ISel but it is definitely easier for the diagnostics with the name itself.

durga4github replied:
Updated to separate intrinsics in the latest revision.

Artem-B replied:
The comment you're referring to is about generating the variant by parametrizing intrinsic definitions in tablegen, not adding parameters to a single intrinsic.

durga4github replied:
ok sure, thanks for clarifying, Artem!

Resolving this since we have moved to separate intrinsics now..

AlexMaclean commented on llvm/include/llvm/IR/IntrinsicsNVVM.td:
```diff
@@ -1466,6 +1466,15 @@ let TargetPrefix = "nvvm" in {
       Intrinsic<[llvm_v2f16_ty], [llvm_i16_ty], [IntrNoMem, IntrNoCallback]>;
 
+// Convert Float to TF32
+def int_nvvm_cvt_float_to_tf32 : Intrinsic<[llvm_i32_ty],
```
Perhaps a name like `llvm.nvvm.convert.to.tf32` would be clearer as it is more consistent with the existing [‘llvm.convert.to.fp16’](https://llvm.org/docs/LangRef.html#id2360) intrinsic.

durga4github replied:
Renamed the intrinsic to "llvm.nvvm.convert.to.tf32.float" in the latest revision

AlexMaclean replied:
Nice, it might be better to suffix it with "f32" instead of "float" so that if we ever support converting from other types we can seamlessly convert to an overloaded intrinsic.

durga4github replied:
Yes, updated to tf32.f32 in the latest revision.

Resolving this,

AlexMaclean commented on llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp:
```diff
@@ -728,7 +728,53 @@ bool NVPTXDAGToDAGISel::tryIntrinsicNoChain(SDNode *N) {
+  }
+}
+
+void NVPTXDAGToDAGISel::SelectCvtFloatToTF32(SDNode *N) {
```
Can this be expressed via table gen patterns?

AlexMaclean replied:
Fair enough, as long as we've considered tablegen this seems fine.

AlexMaclean commented on llvm/include/llvm/IR/IntrinsicsNVVM.td:
```diff
@@ -1466,6 +1466,14 @@ let TargetPrefix = "nvvm" in {
+     llvm_metadata_ty, // Metadata for Rounding modes
+     llvm_i8_ty,       // Flag for Saturation modes
+     llvm_i1_ty],      // Flag for relu
+    [IntrNoMem, ImmArg<ArgIndex<2>>, ImmArg<ArgIndex<3>>]>;
```
Can you add `IntrSpeculatable` to the attrs?

durga4github replied:
Done.

AlexMaclean reviewed (commented):
Thanks for the updates! baring some minor nits this looks good to me. The approaches we adopt here, for representing rounding modes etc., will probably be used in many other future intrinsic so lets make sure we set good standards now. Curious for the thoughts of @Artem-B and @andykaylor on this.

durga4github committed:
[NVPTX] Add convert float to tf32 intrinsics

This patch adds the missing variants of float to tf32
conversion intrinsics. Lit tests are added for all the
intrinsics.

PTX Spec link:
https://docs.nvidia.com/cuda/parallel-thread-execution/#data-movement-and-conversion-instructions-cvt

Signed-off-by: Durgadoss R <durgadossr@nvidia.com>

Artem-B commented on llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:
```diff
@@ -725,6 +728,12 @@ let hasSideEffects = false in {
+  // Float to TF32 conversions.
+  def CVT_tf32_f32 : NVPTXInst<(outs Int32Regs:$dst),
+                     (ins Float32Regs:$src, CvtMode:$mode),
+                     !strconcat("cvt${mode:base}${mode:relu}${mode:satfinite}.",
```
For this limited use case, instead of implementing a separate mode printer, you could just pass the necessary relu/satfinite modifier string for the instruction directly, as the record argument. E.g. instead of `CVT_tf32_f32 $a, CvtRN_RELU` you would use `CVT_tf32_f32 $a, ".relu"` and instead of `CvtRNA_SATFINITE` you'd pass `".rna.satfinite"`.

The extra mode here just complicates things without buying you anything.

Artem-B replied:
To clarify, right now you define a single record which generates all `cvt.tf32.f32` instruction variants, modified via `mode` passed by a custom pattern matcher.

I'm proposing defining per-intrinsic record parametrized by the PTX instruction modifier string, which emits specific PTX instruction, and which applies the intrinsic-matching pattern as a `NVPTXInst` parameter.

durga4github replied:
I have updated to use separate records in the latest revision.

Kindly review it and let me know if it aligns with your suggestion.
